diff --git a/intern/cycles/kernel/film/data_passes.h b/intern/cycles/kernel/film/data_passes.h
index 0105ccdfbb9..acbc4fa0192 100644
--- a/intern/cycles/kernel/film/data_passes.h
+++ b/intern/cycles/kernel/film/data_passes.h
@@ -50,10 +50,6 @@ ccl_device_inline void film_write_data_passes(KernelGlobals kg,
 
   if (!(path_flag & PATH_RAY_SINGLE_PASS_DONE)) {
     if (INTEGRATOR_STATE(state, path, sample) == 0) {
-      if (flag & PASSMASK(DEPTH)) {
-        const float depth = camera_z_depth(kg, sd->P);
-        film_overwrite_pass_float(buffer + kernel_data.film.pass_depth, depth);
-      }
       if (flag & PASSMASK(OBJECT_ID)) {
         const float id = object_pass_id(kg, sd->object);
         film_overwrite_pass_float(buffer + kernel_data.film.pass_object_id, id);
@@ -93,6 +89,13 @@ ccl_device_inline void film_write_data_passes(KernelGlobals kg,
     }
   }
 
+  if (flag & PASSMASK(DEPTH)) {
+    float old_depth = atomic_add_and_fetch_float(buffer + kernel_data.film.pass_depth, 0);
+    while (old_depth == 0.0f) {
+      old_depth = atomic_compare_and_swap_float(buffer + kernel_data.film.pass_depth, old_depth, camera_z_depth(kg, sd->P));
+    }
+  }
+
   if (kernel_data.film.cryptomatte_passes) {
     const Spectrum throughput = INTEGRATOR_STATE(state, path, throughput);
     const float matte_weight = average(throughput) *
diff --git a/source/blender/python/intern/bpy_app.cc b/source/blender/python/intern/bpy_app.cc
index 14c6178bd48..eb3317a5abe 100644
--- a/source/blender/python/intern/bpy_app.cc
+++ b/source/blender/python/intern/bpy_app.cc
@@ -120,6 +120,8 @@ static PyStructSequence_Field app_info_fields[] = {
     /* Modules (not struct sequence). */
     {"icons", "Manage custom icons"},
     {"timers", "Manage timers"},
+
+    {"tsr_depth", "Exists if built with TS1 Renderers depth patch"},
     {nullptr},
 };
 
@@ -207,6 +209,8 @@ static PyObject *make_app_info()
   SetObjItem(BPY_app_icons_module());
   SetObjItem(BPY_app_timers_module());
 
+  SetIntItem(1);
+
 #undef SetIntItem
 #undef SetStrItem
 #undef SetBytesItem
